File: pokerBot.c
#include "pokerBot.h"

// Initializes the PokerBot with a shuffled deck
void initialize_poker_bot(PokerBot *bot) {
    initialize_deck(&bot->deck);
    shuffle_deck(&bot->deck);
}

// Makes a decision (stay or fold) using MCTS
const char *make_decision(PokerBot *bot, Card *hole_cards, Card *community_cards, int community_count) {
    Card initial_state[7];

    // Copy hole cards to initial state
    initial_state[0] = hole_cards[0];
    initial_state[1] = hole_cards[1];

    // Copy community cards
    for (int i = 0; i < community_count; i++) {
        initial_state[2 + i] = community_cards[i];
    }

    // Run MCTS
    MCTSNode *result = mcts_search(initial_state, &bot->deck);

    // Simple decision logic: if win rate > 50%, stay
    if (result->wins > result->visits / 2) {
        return "stay";
    } else {
        return "fold";
    }
}

File: PokerBot.py
from deck import Deck
from hand_evaluator import HandEvaluator
from mcts import MCTS
import random


class PokerBot:
    """
    PokerBot class using Monte Carlo Tree Search (MCTS) to decide fold or stay.
    """

    def __init__(self):
        self.deck = Deck()
        self.evaluator = HandEvaluator()
        self.mcts = MCTS(time_limit=10)

    def make_decision(self, hole_cards, community_cards):
        """
        Makes a decision (fold or stay) based on MCTS win probability.
        """
        known_state = hole_cards + community_cards

        def evaluate_simulation(state):
            simulated_deck = [card for card in self.deck.cards if card not in state]
            random.shuffle(simulated_deck)

            opponent_cards = [simulated_deck.pop(), simulated_deck.pop()]
            full_community = state + simulated_deck[:5 - len(community_cards)]

            bot_score = self.evaluator.evaluate_hand(hole_cards + full_community)
            opponent_score = self.evaluator.evaluate_hand(opponent_cards + full_community)

            return 1 if bot_score > opponent_score else 0

        root = self.mcts.search(known_state, evaluate_simulation)
        win_probability = root.children[0].wins / root.children[0].visits

        return "stay" if win_probability >= 0.5 else "fold"
